File: src/MyServiceBus.Http/HttpTransportFactory.cs
using System.Net;
using System.Collections.Concurrent;
using MyServiceBus.Topology;

namespace MyServiceBus;

public sealed class HttpTransportFactory : ITransportFactory
{
    private readonly ConcurrentDictionary<string, HttpClient> _clients = new();
    private readonly ConcurrentDictionary<Uri, ISendTransport> _sendTransports = new();

    [Throws(typeof(InvalidOperationException), typeof(UriFormatException))]
    public Task<ISendTransport> GetSendTransport(Uri address, CancellationToken cancellationToken = default)
    {
        var baseUri = new Uri(address.GetLeftPart(UriPartial.Authority));
        var client = _clients.GetOrAdd(baseUri.ToString(), _ => new HttpClient { BaseAddress = baseUri });

        var transport = _sendTransports.GetOrAdd(address, uri => new HttpSendTransport(client, uri));
        return Task.FromResult(transport);
    }

    [Throws(typeof(UriFormatException), typeof(PlatformNotSupportedException))]
    public Task<IReceiveTransport> CreateReceiveTransport(
        EndpointDefinition definition,
        Func<ReceiveContext, Task> handler,
        Func<string?, bool>? isMessageTypeRegistered = null,
        CancellationToken cancellationToken = default)
    {
        var uri = new Uri(definition.Address, UriKind.Absolute);
        var listener = new HttpListener();
        var prefix = $"{uri.Scheme}://{uri.Authority}/";
        listener.Prefixes.Add(prefix);

        var transport = new HttpReceiveTransport(
            listener,
            uri.AbsolutePath.Trim('/'),
            handler,
            definition.ConfigureErrorEndpoint,
            isMessageTypeRegistered);

        return Task.FromResult<IReceiveTransport>(transport);
    }
}

File: src/Java/myservicebus-http/src/main/java/com/myservicebus/http/HttpTransportFactory.java
package com.myservicebus.http;

import com.myservicebus.ReceiveTransport;
import com.myservicebus.SendTransport;
import com.myservicebus.TransportFactory;
import com.myservicebus.TransportMessage;
import com.myservicebus.topology.MessageBinding;
import java.net.URI;
import java.net.http.HttpClient;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

public class HttpTransportFactory implements TransportFactory {
    private final ConcurrentHashMap<String, HttpClient> clients = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<URI, HttpSendTransport> transports = new ConcurrentHashMap<>();

    @Override
    public SendTransport getSendTransport(URI address) {
        String key = address.getScheme() + "://" + address.getAuthority();
        HttpClient client = clients.computeIfAbsent(key, k -> HttpClient.newHttpClient());
        return transports.computeIfAbsent(address, uri -> new HttpSendTransport(client, uri));
    }

    @Override
    public ReceiveTransport createReceiveTransport(String address, List<MessageBinding> bindings,
            Function<TransportMessage, CompletableFuture<Void>> handler,
            Function<String, Boolean> isMessageTypeRegistered, int prefetchCount,
            Map<String, Object> queueArguments) throws Exception {
        URI uri = URI.create(address);
        return HttpReceiveTransport.create(uri, handler, true, isMessageTypeRegistered);
    }

    @Override
    public String getPublishAddress(String exchange) {
        return exchange;
    }

    @Override
    public String getSendAddress(String queue) {
        return queue;
    }
}
